### PR TITLE
update transformer init function

### LINKED ISSUE #89829: update transformer init function
### 🚀 The feature, motivation and pitch

Once a transformer instance is created, like below:
`
model_pytorch = nn.Transformer(nhead=16, num_encoder_layers=2, num_decoder_layers=2)
`
I cannot access it's num_encoder_layers by **model_pytorch.num_encoder_layers**, TransformerEncoder, TransformerDecoder, TransformerEncoderLayer, TransformerDecoderLayer alike.  
This is extremely suffering when doing quantization, like below:
`
Transformer.from_torch(model_pytorch)
`
say I have a self-defined Transformer class, it's really hard to create my instance from original pytorch instance, because I cannot  access init value elegantly.

### Alternatives

_No response_

### Additional context

_No response_

cc @jbschlosser @bhosmer @cpuhrsch @erichan1

### PR DESCRIPTION
**Pull request by inisis**
Fixes #89829

cc @jbschlosser @bhosmer @cpuhrsch @erichan1

### CODE REVIEW

**Comment by erichan1** [comment]
This seems reasonable (assuming tests pass). 

An alternative workaround is getting these fields from lower level components. For example, you can find dim_feedforward by going Transformer->TransformerEncoder->feed forward linear layers. What do you think?

File: torch/nn/modules/transformer.py [hunk 1]
```diff
@@ -80,8 +80,16 @@ def __init__(self, d_model: int = 512, nhead: int = 8, num_encoder_layers: int =
...
+        self.num_encoder_layers = num_encoder_layers
+        self.num_decoder_layers = num_decoder_layers
+        self.dim_feedforward = dim_feedforward
+        self.dropout = dropout
+        self.activation = activation
+        self.custom_encoder = custom_encoder
+        self.custom_decoder = custom_decoder
+        self.layer_norm_eps = layer_norm_eps
         self.batch_first = batch_first
+        self.norm_first = norm_first
```
Comment (mikekgfb) [inline torch/nn/modules/transformer.py:92]: This generates actual copies of layers just for bookkeeping, that later use resources for parameters when the model is moved to a GPU.  I'd have strong concerns about keeping modules in the architecture just for "book keeping"

File: torch/nn/modules/transformer.py [hunk 2]
```diff
@@ -80,8 +80,16 @@ def __init__(self, d_model: int = 512, nhead: int = 8, num_encoder_layers: int =
 
         self.d_model = d_model
         self.nhead = nhead
-
+        self.num_encoder_layers = num_encoder_layers
```
Comment (mikekgfb) [inline torch/nn/modules/transformer.py:83]: We often see models changed after the fact.  That raises the question whether we can trust these values after the fact.  Most values can be obtained by introspecting the model.  What would it take to provide methods to do that.?

File: torch/nn/modules/transformer.py [hunk 3]
```diff
@@ -311,6 +320,7 @@ def __init__(self, decoder_layer, num_layers, norm=None):
         super(TransformerDecoder, self).__init__()
         torch._C._log_api_usage_once(f"torch.nn.modules.{self.__class__.__name__}")
         self.layers = _get_clones(decoder_layer, num_layers)
+        self.decoder_layer = decoder_layer
```
Comment (mikekgfb) [inline torch/nn/modules/transformer.py:323]: that's the same as self.layers[0] ?

**Review by mikekgfb — COMMENTED** [review]
This generates actual copies of various modules just for bookkeeping, that later use resources for parameters when the model is moved to a GPU.  I'd have strong concerns about keeping modules in the architecture just for "book keeping"

**Comment by inisis** [comment]
> This seems reasonable (assuming tests pass).
> 
> An alternative workaround is getting these fields from lower level components. For example, you can find dim_feedforward by going Transformer->TransformerEncoder->feed forward linear layers. What do you think?

Hi, this is where the problem lies, you have to dig into your model, and I strongly assume that **Transformer->TransformerEncoder->feed forward linear layers** path not exist, because maybe there is a custom TransformerEncoder in user's Transformer module, or TransformerEncoder has 0 TransformerEncoderLayer. it's messed up.

File: torch/nn/modules/transformer.py (hunk 3 above)
Comment (inisis) [reply torch/nn/modules/transformer.py:323]: Hi, using self.layers[0] may return index error.

File: torch/nn/modules/transformer.py (hunk 1 above)
Comment (inisis) [reply torch/nn/modules/transformer.py:92]: Maybe we can just keep those non-module attributes to save GPU memory, I think those non-module attributes should be saved.

File: torch/nn/modules/transformer.py (hunk 2 above)
Comment (inisis) [reply torch/nn/modules/transformer.py:83]: Maybe we should add a property to get those modules

**Comment by inisis** [comment]
@mikekgfb @erichan1 codes updated in a more elegant way.

**Comment by kit1980** [comment]
@inisis Do you think this may be related to https://github.com/pytorch/pytorch/issues/82443 ?

**Comment by inisis** [comment]
> @inisis Do you think this may be related to #82443 ?

@kit1980 I don't think they are related, and I cannot reproduce your issue on torch version **'1.13.0+cu117'**

**Comment by inisis** [comment]
@mikekgfb @erichan1 can this be merged?

**Pull request closed without merging**